File: src/main/java/com/peasenet/mods/render/ModXray.java
/*
 * Copyright (c) 2022-2023. Gavin Pease and contributors.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
 * associated documentation files (the "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Software, and to permit persons to whom the Software is furnished to do so, subject to the
 * following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
 * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT
 * OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

package com.peasenet.mods.render;

import com.peasenet.gui.mod.xray.GuiXray;
import com.peasenet.main.GavinsMod;
import com.peasenet.mods.Mod;
import com.peasenet.mods.Type;
import com.peasenet.settings.ClickSetting;
import com.peasenet.settings.SubSetting;
import com.peasenet.settings.ToggleSetting;
import com.peasenet.util.RenderUtils;
import com.peasenet.util.event.data.DrawSide;
import com.peasenet.util.listeners.ShouldDrawSideListener;
import net.minecraft.block.BlockState;

/**
 * @author gt3ch1
 * @version 12/31/2022
 * A mod for xray like feature, allowing the player to see through certain blocks.
 */
public class ModXray extends Mod implements ShouldDrawSideListener {

    public ModXray() {
        super(Type.XRAY);
        SubSetting xraySubSetting = new SubSetting(100, 10, getTranslationKey());
        ToggleSetting culling = new ToggleSetting("gavinsmod.settings.xray.culling");
        culling.setCallback(() -> {
            xrayConfig.setBlockCulling(culling.getValue());
            if (isActive()) reload();

        });
        culling.setValue(xrayConfig.shouldCullBlocks());
        ClickSetting menu = new ClickSetting("gavinsmod.settings.xray.blocks");
        menu.setCallback(() -> getClient().setScreen(new GuiXray()));
        xraySubSetting.add(menu);
        xraySubSetting.add(culling);
        addSetting(xraySubSetting);
    }


    /**
     * Checks if a block is visible
     *
     * @param block Block to check
     * @return True if visible, false if not
     */
    public static boolean shouldDrawFace(BlockState block) {
        return xrayConfig.isInList(block.getBlock());
    }

    @Override
    public void onEnable() {
        em.subscribe(ShouldDrawSideListener.class, this);
        super.onEnable();
    }

    @Override
    public void onDisable() {
        em.unsubscribe(ShouldDrawSideListener.class, this);
        super.onDisable();
    }

    @Override
    public void activate() {
        getClient().setChunkCulling(xrayConfig.shouldCullBlocks());
        super.activate();
        reloadRenderer();
    }

    /**
     * Reloads the renderer if and only if the setting "xray.forcereload" is true.
     */
    private void reloadRenderer() {
        getClient().getWorldRenderer().reload();
    }

    @Override
    public void onTick() {
        if (isActive() && !RenderUtils.isHighGamma()) RenderUtils.setHighGamma();
        else if (!GavinsMod.isEnabled(Type.FULL_BRIGHT) && !RenderUtils.isLastGamma() && deactivating) {
            RenderUtils.setLowGamma();
            deactivating = !RenderUtils.isLastGamma();
        }
    }

    @Override
    public boolean isDeactivating() {
        return deactivating;
    }

    @Override
    public void deactivate() {
        // check if full bright is disabled, if it is, reset gamma back to LAST_GAMMA
        if (!GavinsMod.isEnabled(Type.FULL_BRIGHT)) RenderUtils.setLowGamma();
        getClient().setChunkCulling(true);
        reloadRenderer();
        deactivating = true;
        RenderUtils.setGamma(4);
        super.deactivate();
    }

    @Override
    public void onDrawSide(DrawSide event) {
        if (!isActive()) return;
        event.setShouldDraw(shouldDrawFace(event.getState()));
    }
}

File: src/main/kotlin/com/peasenet/mods/render/ModXray.kt
/*
 * MIT License
 *
 * Copyright (c) 2022-2025, Gavin C. Pease
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package com.peasenet.mods.render

import com.peasenet.config.render.XrayConfig
import com.peasenet.gui.mod.render.GuiXraySettings
import com.peasenet.main.Mods
import com.peasenet.main.Settings
import com.peasenet.util.ChatCommand
import com.peasenet.util.RenderUtils
import com.peasenet.util.event.data.BlockEntityRender
import com.peasenet.util.event.data.DrawState
import com.peasenet.util.event.data.TessellateBlock
import com.peasenet.util.listeners.BlockEntityRenderListener
import com.peasenet.util.listeners.ShouldDrawSideListener
import com.peasenet.util.listeners.TessellateBlockListener
import net.minecraft.world.level.block.state.BlockState

/**
 * @author GT3CH1
 * @version 08-16-2026
 * A mod for xray like feature, allowing the player to see through certain blocks.
 */
class ModXray : RenderMod(
    "gavinsmod.mod.render.xray",
    "xray"
), ShouldDrawSideListener, TessellateBlockListener, BlockEntityRenderListener {

    init {
        clickSetting {
            title = translationKey
            callback = {
                client.setScreen(GuiXraySettings())
            }
        }
    }

    override fun onEnable() {
        em.subscribe(ShouldDrawSideListener::class.java, this)
        em.subscribe(TessellateBlockListener::class.java, this)
        em.subscribe(BlockEntityRenderListener::class.java, this)
        super.onEnable()
    }

    override fun onDisable() {
        em.unsubscribe(ShouldDrawSideListener::class.java, this)
        em.unsubscribe(TessellateBlockListener::class.java, this)
        em.unsubscribe(BlockEntityRenderListener::class.java, this)
        super.onDisable()
    }

    override fun activate() {
        client.setChunkCulling(getConfig().blockCulling)
        super.activate()
        client.reloadRenderer()
    }

    override fun onTick() {
        val targetGamma = getConfig().maxGamma().toDouble()
        if (isActive && RenderUtils.gamma < targetGamma) {
            // Persist the player's last-known gamma the first time we boost it.
            if (RenderUtils.gamma <= 1.0) RenderUtils.setHighGamma()
            RenderUtils.gamma = targetGamma
        } else if (!Mods.isActive(ChatCommand.FullBright) && !RenderUtils.isLastGamma && deactivating) {
            RenderUtils.setLowGamma()
            deactivating = !RenderUtils.isLastGamma
        }
    }

    override fun deactivate() {
        if (!Mods.isActive(ChatCommand.FullBright))
            RenderUtils.setLowGamma()
        client.setChunkCulling(true)
        client.reloadRenderer()
        deactivating = true
        super.deactivate()
    }

    override fun onDrawSide(event: DrawState) {
        if (!isActive) return
        event.setShouldDraw(shouldDrawFace(event.state))
    }

    override fun onTessellateBlock(event: TessellateBlock) {
        if (!shouldDrawFace(event.blockState)) event.cancel()
    }

    override fun onRenderBlockEntity(er: BlockEntityRender) {
        if (!shouldDrawFace(client.getWorld().getBlockState(er.entity.blockPos))) er.cancel()
    }

    companion object {
        fun shouldDrawFace(block: BlockState): Boolean {
            return Settings.getConfig<XrayConfig>("xray").isInList(block.block)
        }

        fun getConfig(): XrayConfig = Settings.getConfig("xray")
    }
}
